packages/jsapi-utils/src/MessageUtils.ts

```diff
@@ -0,0 +1,64 @@
+import shortid from 'shortid';
+import Log from '@deephaven/log';
+import { TimeoutError } from '@deephaven/utils';
+
+const log = Log.module('MessageUtils');
+
+export const LOGIN_OPTIONS_REQUEST = 'requestLoginOptionsFromParent';
+export const LOGIN_OPTIONS_RESPONSE = 'loginOptions';
+
+export const SESSION_DETAILS_REQUEST = 'requestSessionDetailsFromParent';
+export const SESSION_DETAILS_RESPONSE = 'sessionDetails';
+
+export interface Message<T> {
+  message: string;
+  payload?: T;
+  id?: string;
+}
+
+export function makeMessage<T>(
+  message: string,
+  id?: string,
+  payload?: T
+): Message<T> {
+  return { id, message, payload };
+}
+
+/**
+ * Request data from the parent window and wait for response
+ * @param request Request message to send to the parent window
+ * @param response Response message to wait for
+ * @param timeout Timeout in ms
+ * @returns Payload of the given type, or undefined
+ */
+export async function requestParentResponse<T>(
+  request: string,
+  response: string,
+  timeout = 30000
+): Promise<T | undefined> {
+  if (window.opener == null) {
+    throw new Error('window.opener is null, unable to send request.');
+  }
+  return new Promise((resolve, reject) => {
+    let timeoutId: NodeJS.Timeout;
+    const id = shortid();
+    const listener = (event: MessageEvent<Message<T>>) => {
+      const { data } = event;
+      log.debug('Received message', data);
+      // TODO: checking just the id should be enough, can probably drop the response message check?
+      if (data?.id !== id || data?.message !== response) {
```

Comment:
Yea I think you're right, if we just look at the `id`, we don't need to bother looking at `response`, and then the signature can just be `requestParentResponse(request, timeout)`... 

Thinking about this a bit further, there's basically three types of messages we can send:
- `request`: Something we expect a `response` for
- `response`: Responding to a `request`
- `notify`: Sending a message we do not expect a response for
Perhaps we should do that instead... or even just have those as the three different message types... `io.deephaven.message.Request`, `io.deephaven.message.Response`, and `io.deephaven.message.Notify`, and then the payload itself has the details...

I'm just trying to think how we'll want to extend this later for multi-window messaging/embedding in iframes. Let's discuss later today.